armi/reactor/converters/uniformMesh.py

```diff
@@ -1130,6 +1160,89 @@ def _calculateReactionRates(lib, keff, assem):
                 continue
             globalFluxInterface.calcReactionRates(b, keff, lib)
 
+    def _calcReactionRatesBlockList(objList, keff, xsNucDict):
+        r"""
+        Compute 1-group reaction rates for the objects in objList (usually a block).
+
+        Parameters
+        ----------
+        objList : List[Block]
+            The list of objects to compute reaction rates on. Notionally this could be upgraded to be
+            any kind of ArmiObject but with params defined as they are it currently is only
+            implemented for a block.
+
+        keff : float
+            The keff of the core. This is required to get the neutron production rate correct
+            via the neutron balance statement (since nuSigF has a 1/keff term).
+
+        xsNucDict: Dict[str, XSNuclide]
+            Microscopic cross sections to use in computing the reaction rates. Keys are
+            nuclide names (e.g., "U235") and values are the associated XSNuclide objects
+            from the cross section library, which contain the microscopic cross section
+            data for a given nuclide in the current cross section group.
+
+        .. impl:: Return the reaction rates for a given ArmiObject
+            :id: I_ARMI_FLUX_RX_RATES_BY_XS_ID
+            :implements: R_ARMI_FLUX_RX_RATES
+
+            This is an alternative implementation of :need:`I_ARMI_FLUX_RX_RATES` that
+            is more efficient when computing reaction rates for a large set of blocks
+            that share a common set of microscopic cross sections.
+
+            For more detail on the reation rate calculations, see :need:`I_ARMI_FLUX_RX_RATES`.
+        """
+        from armi.physics.neutronics.globalFlux.globalFluxInterface import (
+            RX_PARAM_NAMES,
+            RX_ABS_MICRO_LABELS,
+        )
+
+        for obj in objList:
+            rate = collections.defaultdict(float)
+
+            numberDensities = obj.getNumberDensities()
+            try:
+                mgFlux = np.array(obj.getMgFlux())
+            except TypeError:
+                continue
+
+            for nucName, numberDensity in numberDensities.items():
+                if numberDensity == 0.0:
+                    continue
```

Comment:
Instead of looping over `numberDensities` and then checking that the number density is non-zero, would it make more sense to loop over the nuclides in `xsNucDict`? You've already done work to screen out isotopes with zero number density from there, so the number of isotopes in there is less and you don't have to do any further checking for zero values.

I might be misinterpreting this, I'm finding the logic somewhat complex.

Reply:
`xsNucDict` will have an entry for any nuclide that exists at a non-zero number density in any block` in the `blockList`.

This line is checking whether the nuclide has a nonzero density in a particular block or other object.

Reply:
True, maybe the check cannot be removed. But I still think that looping over the other list might make more sense, no? `xsNucDict` is shorter without loss of needed information.

Reply:
`xsNucDict` is not necessarily shorter than `obj.getNumberDensities()`. You could have a list of 500 blocks where 495 of them have 3 nuclides and 5 of them have 400 nuclides, and `xsNucDict` would have 400 nuclides. This is obviously an extreme situation that I've devised here that isn't realistic, but it's just hard to say that one of these lists is always going to be shorter than the other one.

If you refactored, it would be something like:

```
for nucName in xsNucDict:
    numberDensity = numberDensities.get(nucName, 0.0)
    if numberDensity == 0.0:
        continue
```

I think since you always have to get the dict value from `numberDensities`, it makes sense to iterate over `numberDensities.items()` instead of iterating over the keys and then doing `numberDensities.get(key, 0.0)`. In general, in a group of blocks, if _any_ of the blocks have a non-zero number density for a given `nucName`, then you'd have to iterate through that nuclide for every block and run `numberDensities.get(nucName, 0.0)` to check if that nuclide is there.